### ISSUE TITLE
Remove `Ord` and `PartialOrd` from `ObjectReference`

### ISSUE THREAD

**Issue by qinsoon** [mmtk/mmtk-core#884]
We derive `Ord` and `PartialOrd` for `ObjectReference`. The order is determined by the raw address of the reference. However, MMTk core currently does not have any code that uses the ordering of object reference.
https://github.com/mmtk/mmtk-core/blob/5d22c2b3ca6c0f3393f9c8cbd8b47a28868a48cc/src/util/address.rs#L464-L465

`PartialOrd` was added to `ObjectReference` long time ago, on a [commit](https://github.com/mmtk/mmtk-core/commit/0169dc712d1b673f1c7a81ea6140dadada0a75dc#diff-c55d0186f3d5990b5e223c1a8aa1bbe60988201c452dd13b3324e892bc33119e) that introduces clippy lints to our CI tests. Clippy required that. `Ord` was added to `ObjectReference` recently (https://github.com/mmtk/mmtk-core/pull/870). I approved the PR, based on the fact that we already have `PartialOrd` for `ObjectReference`.

I suggest removing `Ord` and `PartialOrd` for `ObjectReference`.

1. No real use case. Currently we do not have a use case where we need to compare the order of two object references. 
2. Unclear meaning for comparing object reference. Object reference represents an object. What does it mean when we say one object reference is less/smaller than or prior to another object reference (https://doc.rust-lang.org/std/cmp/enum.Ordering.html)?
3. Abstraction breakage. We agreed on using an address internally for an object reference. But we do not rule out the possibility that we actually need a completely opaque type as `ObjectReference` in the future. Any code that relies on using the internal address to determine the ordering of object reference may not work if we make `ObjectReference` an opaque type, as the ordering could be different.

I am tagging people who were involved in the discussion in https://github.com/mmtk/mmtk-core/pull/870: @wks @caizixian @k-sareen

**Comment by wks** [comment]
Yes.  It makes sense that the raw address may not be the only possible key to sort ObjectReference.  `objects.sort_by(|obj| obj.to_raw_address())` should be more semantically clear if we do intend to use the raw address.

**Comment by wks** [comment]
But similar arguments can be said for `Hash`, too.  We sometimes need to put `ObjectReference` into `HashMap`.  Currently the `Treadmill` and the `ReferenceProcessor` uses `HashSet` to test if a given `ObjectReference` is in the set.  The `Hash` trait is also implemented by hashing the raw address, but we don't consider it as a leak of abstraction, given that

1.  Each valid object has a unique `ObjectReference` raw address determined by `ObjectModel::address_to_ref`, and
2.  two `ObjectReference` instances are equal if and only if they point to the same object, which is equivalent to having the same raw address.

The only requirement is that two `ObjectReference` instances that compare equal should have the same hash code, and it is better if different `ObjectReference` instances have different hash code.

It is similar for ordered containers such as `BTreeSet` and `BTreeMap`.  It requires a total order for the value.  This total order may not be the natural ordering of the keys (if exists at all).  For example, in Java, `Double#compareTo(Double)` implements a total order so that `Double` can be used in ordered containers.  But IEEE754 floating numbers is not totally ordered due to the presence of NaN. (See [JavaDoc](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/lang/Double.html#compareTo(java.lang.Double))

Well, Rust does **not** implement `Ord` for `f32` because the natural total order doesn't exist, and it cannot be put into an ordered container, either.  And `f32` doesn't implement `Hash`, either.

To answer some of @qinsoon's concerns

>   2.  Unclear meaning for comparing object reference

`Ord` provides a total ordering for `ObjectReference`, although it is based on a hidden internal value.  It's just like the hash function depends on the same value.

>   3.  Abstraction leakage

Currently, we know that `ObjectReference` uniquely maps to an object, and it is comparable.  Adding a total order (without revealing what that order is based on) doesn't leak more details to the public.

So a real problem is that we are not making use of the total ordering for now.  If we want to allow `ObjectReference` to be put into ordered containers such as `BTreeSet`, we can keep the `Ord` trait.  But from the example of `f32`, we can still choose not to implement `Ord` because it feels unnatural to implement `Ord` for something that doesn't have a natural order.  Even if we don't do that, if the users really want to, they can still put `ObjectOrdering` into `BTreeSet` using a classic trick used by Haskell:

```rs
#[derive(PartialEq, Eq, PartialOrd)]
struct ByRawAddress(ObjectReference);
impl std::cmp::Ord for ByRawAddress {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.0.to_raw_address().cmp(&other.0.to_raw_address())
    }
}

let hs = HashSet::<ByRawAddress>::new();
```

When ordering a list for accessing metadata via raw addresses, I think it's better to stick to `objects.sort_by(|obj| obj.to_raw_address())`.
